test/adt_test.exs

```diff
@@ -54,6 +54,29 @@ defmodule AdtTest do
     assert AdtDefinitionThree.variants == [AdtTest.AdtDefinitionThree.Baz, AdtTest.AdtDefinitionThree.Bar, AdtTest.AdtDefinitionThree.Foo]
   end
 
+  test "case" do
+    require AdtTest.AdtDefinition
+
+    foo = %AdtDefinition.Foo{}
+
+    result = AdtDefinition.case foo, [
+      Foo: fn(_) -> "1" end,
+      Bar: fn(_) -> "2" end
+    ]
+    assert result == "1"
+  end
+
+  test "non-exhaustive case" do
+    error = catch_error Code.eval_string """
+      require AdtTest.AdtDefinition
+
+      result = AdtTest.AdtDefinition.case %AdtTest.AdtDefinition.Foo{}, [
+        Foo: fn(x) -> "foo" end
+      ]
+    """
+    assert error.message == "case macro not exhaustive.\nGiven [\"Foo\"].\nPossible: [\"Bar\", \"Foo\"]."
```

Comment:
It's definitely worth thinking about, one thing I'd like to cover is dead code. Given:

```elixir
  defmodule AdtDefinition do
    ADT.define foo(a: 0) | bar(val: "hey")
  end
```

```elixir
    result = AdtDefinition.case val, [
      Foo: fn(_) -> "1" end,
      Bar: fn(_) -> "2" end,
      Baz: fn(_) -> "3" end
    ]
```

I think this should _also_ raise, as the programmer has written dead code in the `Baz` case.